org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java

```diff
@@ -0,0 +1,208 @@
+			if (node.isAttribute() && ((DOMAttr) node).getOwnerElement() != null) {
+				node = ((DOMAttr) node).getOwnerElement();
+			}
+			DOMNode prolog = node.getOwnerDocument().getProlog();
+			if (prolog != null && node != node.getOwnerDocument() && node != prolog && prolog.getStart() <= offset
+					&& offset <= prolog.getEnd()) {
+				node = node.getOwnerDocument().getProlog();
+			} else {
+				node = node.getParentNode();
+			}
+		} while (node != null);
+
+		// Skip the leaf node, which is a dummy node
+		return leafSelectionRange.getParent();
+	}
+
+	private SelectionRange handleNodeSelectionRange(SelectionRange selectionRange, DOMNode node, int offset) {
+		switch (node.getNodeType()) {
+			case Node.ELEMENT_NODE:
+				return handleElementSelectionRange(selectionRange, (DOMElement) node, offset);
+			case Node.ATTRIBUTE_NODE:
+				return handleAttributeSelectionRange(selectionRange, (DOMAttr) node, offset);
+			case Node.DOCUMENT_TYPE_NODE:
+				return handleDoctypeSelectionRange(selectionRange, (DOMDocumentType) node, offset);
+			default:
+				return handleGenericNodeSelectionRange(selectionRange, node, offset);
+		}
+	}
+
+	private SelectionRange handleAttributeSelectionRange(SelectionRange selectionRange, DOMAttr attr, int offset) {
+		selectionRange = handleGenericNodeSelectionRange(selectionRange, attr, offset);
+		DOMElement ownerElement = attr.getOwnerElement();
+		if (ownerElement != null) {
+			selectionRange = handleGenericNodeSelectionRange(selectionRange, ownerElement, offset);
+		}
+		return selectionRange;
+	}
+
+	private SelectionRange handleElementSelectionRange(SelectionRange selectionRange, DOMElement element, int offset) {
+		XMLPositionUtility.selectStartTagName(element);
+		int contentStart = element.getStartTagCloseOffset() + 1;
+		int contentEnd = element.getEndTagOpenOffset();
+		if (contentStart != DOMNode.NULL_VALUE && contentEnd != DOMNode.NULL_VALUE && offset >= contentStart
+				&& offset <= contentEnd) {
+			Range range = XMLPositionUtility.createRange(contentStart, contentEnd, element.getOwnerDocument());
+			if (range != null && !range.equals(selectionRange.getRange())) {
+				SelectionRange parentSelectionRange = new SelectionRange();
+				parentSelectionRange.setRange(range);
+				selectionRange.setParent(parentSelectionRange);
+				selectionRange = parentSelectionRange;
+			}
+		} else if (contentStart != DOMNode.NULL_VALUE && offset < contentStart) {
+			SelectionRange parentSelectionRange = new SelectionRange();
+			parentSelectionRange.setRange(XMLPositionUtility.selectStartTagName(element));
+			selectionRange.setParent(parentSelectionRange);
+			selectionRange = parentSelectionRange;
+		} else if (offset <= element.getEndTagCloseOffset()) {
+			SelectionRange parentSelectionRange = new SelectionRange();
+			parentSelectionRange.setRange(XMLPositionUtility.selectEndTagName(element));
+			selectionRange.setParent(parentSelectionRange);
+			selectionRange = parentSelectionRange;
```

Comment:
Would it make more sense to move :

`SelectionRange parentSelectionRange = new SelectionRange();`

above the start of the if statement and :

`selectionRange.setParent(parentSelectionRange);`
`selectionRange = parentSelectionRange;`

below the if block ?

Those statements seem to stay the same.

This would leave just `parentSelectionRange.setRange(...)` within the conditionals.

Reply:
There are some cases where I want to avoid adding a new parent selection range, for instance when the cursor is here:
```xml
...
  |<element>content</element>
...
```
I just want the range that covers the entire element, so I don't add a step before selecting the entire element.

I moved the `parentSelectionRange` creation to the beginning, since it can be GC'd and it makes the code easier to read